### PR TITLE
feat: Iteration node support parallel mode

### LINKED ISSUE #8226: Add Parallel and Loop Module Functionality to Iteration Block + Support Donation Feature
### Self Checks

- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] [FOR CHINESE USERS] 请务必使用英文提交 Issue，否则会被关闭。谢谢！:）
- [X] Please do not modify this template :) and fill in all the required fields.

### 1. Is this request related to a challenge you're experiencing? Tell me about your story.

I upgraded to V0.8.0 immediately, and I really appreciate the new parallel processing feature. It's a fantastic improvement—thank you for your hard work! However, I noticed that while the parallel processing works well in general, it doesn’t currently apply to workflows within the iteration module.

My suggestion is to add an option in the iteration module that allows users to toggle between two modes: parallel processing and the default linear processing (as it was before). In parallel mode, the iteration module could distribute the array of information received from the previous node and spawn multiple tasks that can be processed simultaneously. This could further speed up the execution of complex workflows, making them more efficient.

Additionally, I would suggest adding a "Loop" module where users can set conditions for exiting the loop. As long as the condition isn’t met, the workflow inside the loop would keep running. Once the condition is met, it would proceed to the next node. This would be a basic yet powerful tool to expand the options available when building workflows and enable more complex functionality.

### 2. Additional context or comments

Lastly, I would like to suggest adding a way to support the project, such as a "Buy me a coffee" option. I believe that a sustainable financial model will help the project thrive in the long term. If a commercial model hasn’t been decided yet, this could allow us, the users, to make small contributions to support its development.

### 3. Can you help us with this feature?

- [X] I am interested in contributing to this feature.

### PR DESCRIPTION
**Pull request by Nov1c444**
# Checklist:

> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.

- [x] Please open an issue before creating a PR or link to an existing issue
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

# Description

Iteration Node supports parallel mode. And add three error handle mode. 

Fixes #8226 #9121

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [ ] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# Testing Instructions

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

### CODE REVIEW

**Review by iamjoel — DISMISSED** [review]

File: api/core/workflow/nodes/iteration/iteration_node.py [hunk 1]
```diff
@@ -330,3 +308,230 @@ def _extract_variable_selector_to_variable_mapping(
...
+        return event
+
+    def _run_single_iter(
+        self,
+        iterator_list_value: list[str],
+        variable_pool: VariablePool,
+        inputs: dict[str, list],
+        outputs: list,
+        start_at: datetime,
+        graph_engine,
```
Comment (laipz8200) [inline api/core/workflow/nodes/iteration/iteration_node.py]: please add type hints here.

File: api/core/workflow/nodes/iteration/iteration_node.py [hunk 2]
```diff
@@ -330,3 +308,230 @@ def _extract_variable_selector_to_variable_mapping(
...
+        iteration_graph,
```
Comment (laipz8200) [inline api/core/workflow/nodes/iteration/iteration_node.py]: please add type hints here.

File: api/core/workflow/nodes/iteration/iteration_node.py [hunk 3]
```diff
@@ -330,3 +308,230 @@ def _extract_variable_selector_to_variable_mapping(
...
+
+    def _run_single_iter_parallel(
+        self,
+        flask_app: Flask,
+        q: Queue,
+        iterator_list_value: list[str],
+        inputs: dict[str, list],
+        outputs: list,
+        start_at: datetime,
+        graph_engine,
```
Comment (laipz8200) [inline api/core/workflow/nodes/iteration/iteration_node.py]: please add type hints here.

File: api/core/workflow/nodes/iteration/iteration_node.py [hunk 4]
```diff
@@ -330,3 +308,230 @@ def _extract_variable_selector_to_variable_mapping(
...
+        iteration_graph,
```
Comment (laipz8200) [inline api/core/workflow/nodes/iteration/iteration_node.py]: please add type hints here.

File: api/core/workflow/nodes/iteration/iteration_node.py [hunk 5]
```diff
@@ -330,3 +308,230 @@ def _extract_variable_selector_to_variable_mapping(
...
+        index,
```
Comment (laipz8200) [inline api/core/workflow/nodes/iteration/iteration_node.py]: please add type hints here.

File: api/core/workflow/nodes/iteration/iteration_node.py [hunk 6]
```diff
@@ -330,3 +308,230 @@ def _extract_variable_selector_to_variable_mapping(
...
+        item,
```
Comment (laipz8200) [inline api/core/workflow/nodes/iteration/iteration_node.py]: please add type hints here.

File: api/core/workflow/nodes/iteration/entities.py [hunk 7]
```diff
@@ -1,10 +1,24 @@
+from enum import Enum
 from typing import Any, Optional
 
 from pydantic import Field
 
 from core.workflow.nodes.base import BaseIterationNodeData, BaseIterationState, BaseNodeData
 
 
+class ErrorHandleMode(Enum):
```
Comment (laipz8200) [inline api/core/workflow/nodes/iteration/entities.py]: I think this enum should inherit from `str`.
```suggestion
class ErrorHandleMode(str, Enum):
```

File: api/core/workflow/nodes/iteration/entities.py [hunk 8]
```diff
@@ -1,10 +1,24 @@
...
+    TERMINATED = "Terminated"
+    CONTINUE_ON_ERROR = "Continue on error"
+    REMOVE_ABNORMAL_OUTPUT = "Remove abnormal output"
+
+    def to_json(self):
```
Comment (laipz8200) [inline api/core/workflow/nodes/iteration/entities.py]: What is this method used for?

File: api/core/workflow/nodes/iteration/entities.py [hunk 9]
```diff
@@ -1,10 +1,24 @@
...
+        return self.value
+
+    @classmethod
+    def from_json(cls, value):
```
Comment (laipz8200) [inline api/core/workflow/nodes/iteration/entities.py]: What is this method used for?

**Review by laipz8200 — CHANGES_REQUESTED** [review]

File: api/core/workflow/nodes/iteration/entities.py [hunk 10]
```diff
@@ -1,10 +1,17 @@
...
 
 from pydantic import Field
 
 from core.workflow.nodes.base import BaseIterationNodeData, BaseIterationState, BaseNodeData
 
 
+class ErrorHandleMode(str, Enum):
+    TERMINATED = "Terminated"
+    CONTINUE_ON_ERROR = "Continue on error"
+    REMOVE_ABNORMAL_OUTPUT = "Remove abnormal output"
```
Comment (laipz8200) [inline api/core/workflow/nodes/iteration/entities.py]: The enumeration here can use lowercase letters + hyphen (-), just like in other places.

**Review by laipz8200 — DISMISSED** [review]

**Review by laipz8200 — APPROVED** [review]

**Comment by detongz** [comment]
Iteration node parallel mode is great, but I find that it wont work when i have parallel llm nodes running within it:
image
image

After iteration node parallel mode the entire loop blocks at its first node and stucked.
image

**Comment by Nov1c444** [comment]
> Iteration node parallel mode is great, but I find that it wont work when i have parallel llm nodes running within it: image image
> 
> After iteration node parallel mode the entire loop blocks at its first node and stucked. image
Sorry, I tried to build a similar workflow, but couldn't reproduce your issue. Could you please create an issue and provide us with more information?

**Pull request merged**